### PR TITLE
Minion: Add callbacks for solutions fetching

### PR DESCRIPTION
**Pull request by niklasdewally**
This PR adds callbacks to programatically fetch solutions from Minion into Rust.

Also has misc README, licensing, and documentation changes.

### CODE REVIEW

**Comment by niklasdewally** [comment]
@ozgurakgun Could I have a review?

**Comment by niklasdewally** [comment]
Doing some documentation, will remark as ready when this is done.

File: solvers/minion/src/run.rs [hunk 1]
```diff
@@ -1,29 +1,165 @@
...
+///   // Get solutions
+///   let guard = ALL_SOLUTIONS.lock().unwrap();
+///
+///   let solution_set_1 = &(*guard.get(0).unwrap());
+///
+///   let x1 = solution_set_1.get("x").unwrap();
+///   let y1 = solution_set_1.get("y").unwrap();
+///   let z1 = solution_set_1.get("z").unwrap();
+/// #
+/// # // FIXME: this test would be better with an example with >1 solution.
```
Comment (ozgurakgun) [inline solvers/minion/src/run.rs:92]: just make one of the domains larger?

**Review by ozgurakgun — COMMENTED** [review]
I just took a quick pass and I like what I see! I think this is ready to be merged and we can then focus on getting some Essence->AstJson->Minion pipeline.

File: solvers/minion/src/run.rs [hunk 2]
```diff
@@ -1,29 +1,165 @@
...
 #[no_mangle]
-extern "C" fn hello_from_rust2() -> bool {
-    return true;
+unsafe extern "C" fn run_callback() -> bool {
+    // get printvars from static PRINT_VARS if they exist.
+    // if not, return true and continue search.
+
+    let mut guard: MutexGuard<'_, Option<Vec<VarName>>> = PRINT_VARS.lock().unwrap();
+
+    println!("here 0");
```
Comment (niklasdewally) [inline solvers/minion/src/run.rs:116]: TODO Remove

File: solvers/minion/src/run.rs [hunk 3]
```diff
@@ -1,29 +1,165 @@
...
+    if guard.is_none() {
+        return true;
+    }
+
+    let print_vars = match &mut *guard {
+        Some(x) => x,
+        None => unreachable!(),
+    };
+
+    println!("here 1 printvars: {:?}", print_vars);
```
Comment (niklasdewally) [inline solvers/minion/src/run.rs:126]: TODO Remove

File: solvers/minion/src/run.rs [hunk 4]
```diff
@@ -1,29 +1,165 @@
...
+    // build nice solutions view to be used by callback
+    let mut solutions: HashMap<VarName, Constant> = HashMap::new();
+
+    for (i, var) in print_vars.iter().enumerate() {
+        let solution_int: i32 = printMatrix_getValue(i as _).try_into().unwrap();
+        let solution: Constant = Constant::Integer(solution_int);
+        solutions.insert(var.to_string(), solution);
+    }
+
+    println!("here {:?}", solutions);
```
Comment (niklasdewally) [inline solvers/minion/src/run.rs:140]: TODO Remove

**Review by ChrisJefferson — APPROVED** [review]

**Comment by ozgurakgun** [comment]
Updating this branch just in case the new dependency updates break the build (unlikely)

File: solvers/minion/src/run.rs (hunk 1 above)
Comment (niklasdewally) [reply solvers/minion/src/run.rs]: I think I have Minion set to "one solution only". Also, as discussed on Teams, I am going to rethink how I am going to do examples generally

**Comment by niklasdewally** [comment]
@ozgurakgun I think this is ready to go.

I have not refined that example, but, as discussed on Teams, I want to rethink how I write documentation examples anyways, so it is likely to disappear soon.
I think this is better done in a later PR so that this code can be used to get a pipeline going.

**Comment by ozgurakgun** [comment]
Do you know why one of the chuffed builds fail? I restarted it but it failed again.

**Comment by niklasdewally** [comment]
> Do you know why one of the chuffed builds fail? I restarted it but it failed again.

Also happens in #57 - seems to be a macos / rust version change has broke things.

Rustc segfaults and crashes, so I doubt its something we did!

**Pull request merged**